job/service.go

```diff
@@ -738,3 +692,194 @@ func populateDownstreamDAGs(dagTree *tree.MultiRootTree, jobSpec models.JobSpec,
+		runner.Add(func(currentSpec models.JobSpec) func() (interface{}, error) {
+			return func() (interface{}, error) {
+				err := srv.dependencyResolver.ResolveAndPersist(ctx, projectSpec, currentSpec, progressObserver)
+				if err != nil {
+					return currentSpec.Name, fmt.Errorf("%s: %s: %w", errDependencyResolution, currentSpec.Name, err)
+				}
+				return currentSpec.Name, nil
+			}
+		}(jobSpec))
+	}
+
+	failure, succeed := 0, 0
+	for _, state := range runner.Run() {
+		if state.Err != nil {
+			err = multierror.Append(err, state.Err)
+			failure++
+			srv.notifyProgress(progressObserver, &models.ProgressJobDependencyResolution{Job: fmt.Sprintf("%v", state.Val), Err: state.Err})
+		} else {
+			succeed++
+			srv.notifyProgress(progressObserver, &models.ProgressJobDependencyResolution{Job: fmt.Sprintf("%v", state.Val)})
+		}
+	}
+
+	resolveDependencySuccessGauge.Set(float64(succeed))
+	resolveDependencyFailureGauge.Set(float64(failure))
+	srv.notifyProgress(progressObserver, &models.ProgressJobDependencyResolutionFinished{})
+	return err
+}
+
+func prepareNamespaceJobsMap(ctx context.Context, projectJobSpecRepo store.ProjectJobSpecRepository,
+	jobSpecs []models.JobSpec) (map[string][]models.JobSpec, error) {
+	// Group per namespace
+	jobsToNamespace, err := prepareJobToNamespaceMap(ctx, projectJobSpecRepo)
+	if err != nil {
+		return nil, err
+	}
+
+	namespaceJobSpecMap := make(map[string][]models.JobSpec)
+	for _, jobSpec := range jobSpecs {
+		namespaceJobSpecMap[jobsToNamespace[jobSpec.Name]] = append(namespaceJobSpecMap[jobsToNamespace[jobSpec.Name]], jobSpec)
+	}
+	return namespaceJobSpecMap, nil
+}
+
+func prepareJobToNamespaceMap(ctx context.Context, projectJobSpecRepo store.ProjectJobSpecRepository) (map[string]string, error) {
+	namespaceToJobs, err := projectJobSpecRepo.GetJobNamespaces(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve namespace to job mapping: %w", err)
+	}
+	// generate a reverse map for namespace
+	jobsToNamespace := map[string]string{}
+	for ns, jobNames := range namespaceToJobs {
+		for _, jobName := range jobNames {
+			jobsToNamespace[jobName] = ns
+		}
+	}
+	return jobsToNamespace, nil
+}
+
+func (srv *Service) prepareAllJobsInProject(ctx context.Context, jobDependencies map[uuid.UUID][]models.JobSpecDependency,
+	projectJobSpecRepo store.ProjectJobSpecRepository) ([]models.JobSpec, error) {
+	jobSpecs, err := projectJobSpecRepo.GetAll(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// enrich job dependencies
+	for _, jobSpec := range jobSpecs {
+		dependencies := jobDependencies[jobSpec.ID]
+		for _, dependency := range dependencies {
+			jobSpec.Dependencies[dependency.Job.Name] = dependency
+		}
+	}
+
+	// enrich hook dependencies
+	for _, jobSpec := range jobSpecs {
+		jobSpec.Hooks = srv.dependencyResolver.FetchHookWithDependencies(jobSpec)
+	}
+
+	return jobSpecs, nil
+}
+
+func (srv *Service) prepareJobSpecs(ctx context.Context, projectSpec models.ProjectSpec,
```

Comment:
prepare is generic better to have specific functions so readers can easily understand whats happening

Reply:
renamed this to `fetchJobSpecs`